### ISSUE TITLE
sqlsmith bug hunt: channels not propagating errors

### ISSUE THREAD

**Issue by kwannoel** [risingwavelabs/risingwave#3908]
Some expressions might only fail during runtime, for instance if they only fetch columns.
Errors should be propagated. In the following cases, the underlying error is `NumericOutOfRangeError`, but it is not propagated to downstream executors.

```
dev=> create table t(v int);                                                                                                                                                                                                    CREATE_TABLE
dev=> insert into t values (33);                                                                                                                                                                                                             INSERT 0 1
dev=> select (SMALLINT '22' << v) from t;                                                                                                                                                                                                    ERROR:  RPC error: Status { code: Internal, message: "internal error: broken fifo_channel", metadata: MetadataMap { headers: {"risingwave-error-bin": "CAESI2ludGVybmFsIGVycm9yOiBicm9rZW4gZmlmb19jaGFubmVs"} }, source: None }
```

```
dev=> create table t(v int);                                                                                                                                                                                                    CREATE_TABLE
dev=> insert into t values (33);                                                                                                                                                                                                             INSERT 0 1
dev=> select (SMALLINT '22' << v) from t;                                                                                                                                                                                                    ERROR:  RPC error: Status { code: Internal, message: "internal error: broken fifo_channel", metadata: MetadataMap { headers: {"risingwave-error-bin": "CAESI2ludGVybmFsIGVycm9yOiBicm9rZW4gZmlmb19jaGFubmVs"} }, source: None }
dev => insert into t values(0);
INSERT 0 1
dev=> select (SMALLINT '33' / SMALLINT v) from t;
ERROR:  RPC error: Status { code: Internal, message: "internal error: broken fifo_channel", metadata: MetadataMap { headers: {"risingwave-error-bin": "CAESI2ludGVybmFsIGVycm9yOiBicm9rZW4gZmlmb19jaGFubmVs"} }, source: None }

```

**Comment by kwannoel** [comment]
This is because the channel we use between executors passes `Optionals` around:
https://github.com/singularity-data/risingwave/blob/b64d8294c091413fa517eb70a0659ae3c6bcd3d0/src/batch/src/task/fifo_channel.rs#L31

And so when we read from the channel, the errors are not propagated:
https://github.com/singularity-data/risingwave/blob/b64d8294c091413fa517eb70a0659ae3c6bcd3d0/src/batch/src/task/fifo_channel.rs#L47-L59

Am thinking of changing the Item type in fifo channel to `Result` for error propagation. Any other suggestions?

**Comment by skyzh** [comment]
`Result` might not be serializable and cannot be sent over the network. I think the current implementation is okay. If there's an error, we can find it on upstream compute node?

We can modify the error message to be more user friendly, like "internal error when computing" [E2333], where E2333 = fifo channel broken. Then developers can go into the log dashboard (maybe ELK in the future) to find what's happening.

**Comment by kwannoel** [comment]
> Result might not be serializable and cannot be sent over the network. I think the current implementation is okay. If there's an error, we can find it on upstream compute node?
> 
> We can modify the error message to be more user friendly, like "[E2333] internal error when computing", where E2333 = fifo channel broken. Then developers can go into the log dashboard (maybe ELK in the future) to find what's happening.

That makes sense. I guess our fuzzers need to generate sql w/o undefined behaviour, since those may cause opaque internal errors as above. Will track that in separate issue.

**Comment by kwannoel** [comment]
Actually what about other defined errors? E.g. division by zero. These errors are also not propagated too. cc @skyzh . :thinking:

**Comment by Graphcalibur** [comment]
Related issue: #2473

**Comment by kwannoel** [comment]
Had an offline discussion with @lmatz and he suggested the following:
Multiple upstream executors may yield multiple errors, causing the errors to accumulate. Does not make sense to pass all these downstream, it becomes exponentially larger.
One way to deal with this is by having a separate mechanism for tracking all errors.
We can push all errors to some error queue as they are encountered, then frontend can use this to view errors.

This currently blocks sqlsmith work, since we can't know what the source of `broken fifo channel` error is: Runtime error? System internal error? etc...

This issue will take a fair amount of work to resolve.
As a workaround to unblock SqlSmith work, we let `broken fifo channel` errors pass through for now.

**Comment by xxchan** [comment]
related issue https://github.com/singularity-data/risingwave/issues/3730

**Comment by BowenXiao1999** [comment]
Create a issue relates to this: https://github.com/singularity-data/risingwave/issues/4811
